nexus-sdk/src/main/java/io/nexusrpc/handler/ServiceHandler.java

```diff
@@ -0,0 +1,203 @@
+package io.nexusrpc.handler;
+
+import io.nexusrpc.*;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.util.*;
+import org.jspecify.annotations.Nullable;
+
+/** Handler that delegates to service implementations. */
+public class ServiceHandler implements Handler {
+  /** Create a builder for a service handler. */
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  /** Create a builder for service handler from an existing service handler. */
+  public static Builder newBuilder(ServiceHandler handler) {
+    return new Builder(handler);
+  }
+
+  private final Map<String, ServiceImplInstance> instances;
+  private final Serializer serializer;
+
+  private ServiceHandler(Map<String, ServiceImplInstance> instances, Serializer serializer) {
+    this.instances = instances;
+    this.serializer = serializer;
+  }
+
+  /** Instances, by service name. */
+  public Map<String, ServiceImplInstance> getInstances() {
+    return instances;
+  }
+
+  /** Serializer used for input/output. */
+  public Serializer getSerializer() {
+    return serializer;
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public OperationStartResult<HandlerResultContent> startOperation(
+      OperationContext context, OperationStartDetails details, InputStream param)
+      throws UnrecognizedOperationException, OperationUnsuccessfulException {
+    ServiceImplInstance instance = instances.get(context.getService());
+    if (instance == null) {
+      throw new UnrecognizedOperationException(context.getService(), context.getOperation());
+    }
+    OperationHandler<Object, Object> handler =
+        instance.getOperationHandlers().get(context.getOperation());
+    if (handler == null) {
+      throw new UnrecognizedOperationException(context.getService(), context.getOperation());
+    }
+    OperationDefinition definition =
+        instance.getDefinition().getOperations().get(context.getOperation());
+
+    Object input;
+    try {
+      // Collect entire input stream as byte array. This is unfortunately the best Java-8-safe
+      // approach.
+      ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+      int nRead;
+      byte[] data = new byte[1024];
+      while ((nRead = param.read(data, 0, data.length)) != -1) {
+        buffer.write(data, 0, nRead);
+      }
+
+      // Deserialize to expected input type
+      Serializer.Content.Builder contentBuilder = Serializer.Content.newBuilder();
+      contentBuilder.setData(buffer.toByteArray());
+      contentBuilder.getHeaders().putAll(context.getHeaders());
```

Comment:
I think I got this part wrong. I think I need to accept a `Content` instead of just taking all headers from the operation to the content.

Reply:
Fixed.